scripts/ghprb_deduper.py

```diff
@@ -0,0 +1,237 @@
+"""
+Requires:
+    requests==2.5.3
+"""
+from collections import defaultdict
+from operator import itemgetter
+import argparse
+import json
+import re
+import requests
+import logging
+logging.basicConfig(format='[%(levelname)s] %(message)s')
+logger = logging.getLogger(__name__)
+
+
+class GhprbOutdatedBuildAborter():
+    """
+    A class for programatically finding and aborting outdated
+    jenkins builds that were started using the GHPRB plugin.
+
+    :Args:
+        job_url: URL of jenkins job that uses the GHPRB plugin
+        username: jenkins username
+        token: jeknins api token
+    """
+    def __init__(self, job_url, username, token):
+        self.job_url = job_url
+        self.auth = (username, token)
+
+    def _append_url(self, base, addition):
+        """
+        Add something to a url, ensuring that there are the
+        right amount of `/`.
+
+        :Args:
+            base: The original url.
+            addition: the thing to add to the end of the url
+
+        :Returns: The combined url as a string of the form
+            `base/addition`
+        """
+        if not base.endswith('/'):
+            base += '/'
+        if addition.startswith('/'):
+            addition = addition[1:]
+        return base + addition
+
+    def _aborted_description(self, current_build_id, pr):
+        """
+        :Args:
+            current_build_id: the id of the most recently started
+                build for the PR
+            pr: the PR id
+
+        :Returns: A description (string)
+        """
+        return ("[PR #{}] Build automatically aborted because"
+                " there is a newer build for the same PR. See build"
+                " #{}.".format(pr, current_build_id))
+
+    def abort_duplicate_builds(self):
+        """
+        Find running builds of the ghprb job at self.job_url.
+        If there are multiple jobs running for the same PR,
+        abort all but the newest build. It updates the build
+        description of aborted jobs to indicate why they where
+        stopped.
+        """
+        data = self.get_json()
+        builds = self.get_running_builds(data)
+        self.stop_duplicates(builds)
+
+    def get_json(self):
+        """
+        Get build data for a given job_url.
+
+        :Returns:
+            A python dict from the jeknins api response including:
+            * builds: a list of dicts, each containing:
+                ** building: Boolean of whether it is actively building
+                ** timestamp: the time the build started
+                ** number: the build id number
+                ** actions: a list of 'actions', from which the only
+                    item used in this script is 'parameters' which can
+                    be used to find the PR number.
+        """
+        api_url = self._append_url(self.job_url, '/api/json')
+
+        response = requests.get(
```

Comment:
As written, the rest of the script will just fail because there is no data to parse.  Same for the case where a wrong url is passed. If it is a wrong url that happens to be using ghprb and the user has permissions to it, then it will all work...  It's relying on the user to input correct things. 

I would probably opt for using `response.raise_for_status()` instead of a try/except here.